ads/alp/handler.js

```diff
@@ -0,0 +1,182 @@
+ * @visibleForTesting
+ */
+export function handleClick(e) {
+  if (e.defaultPrevented) {
+    return;
+  }
+  // Only handle simple clicks with the left mouse button/touch and without
+  // modifier keys.
+  if (e.buttons != 0 && e.buttons != 1) {
+    return;
+  }
+  if (e.ctrlKey || e.altKey || e.shiftKey || e.metaKey) {
+    return;
+  }
+
+  const link = getLinkInfo(e);
+  if (!link || !link.eventualUrl) {
+    return;
+  }
+
+  // Tag the original href with &amp=1 and make it a fragment param with
+  // name click.
+  const fragment = 'click=' + encodeURIComponent(
+      addParamToUrl(link.a.href, 'amp', '1'));
+  let destination = link.eventualUrl;
+  if (link.eventualUrl.indexOf('#') == -1) {
+    destination += '#' + fragment;
+  } else {
+    destination += '&' + fragment;
+  }
+  const win = link.a.ownerDocument.defaultView;
+  const ancestors = win.location.ancestorOrigins;
+  if (ancestors && ancestors[ancestors.length - 1] == 'http://localhost:8000') {
+    destination = destination.replace('https://cdn.ampproject.org/c/',
+        'http://localhost:8000/max/');
+  }
+
+  e.preventDefault();
+  navigateTo(win, link.a, destination);
+}
+
+/**
+ * For an event, see if there is an anchor tag in the target
+ * ancestor chain and if yes, check whether we can figure
+ * out an AMP target URL.
+ * @param {!Event} e
+ * @return {{
+ *   eventualUrl: (string|undefined),
+ *   a: !Element
+ * }} A URL on the AMP Cache.
+ */
+function getLinkInfo(e) {
+  const a = closest(e.target, element => {
+    return element.tagName == 'A' && element.href;
+  });
+  if (!a) {
+    return;
+  }
+  return {
+    eventualUrl: getEventualUrl(a),
+    a,
+  };
+}
+
+/**
+ * Given an anchor tag, figure out whether this goes to an AMP destination
+ * via a redirect.
+ * @param {!Element} a An anchor tag.
+ * @return {string|undefined} A URL on the AMP Cache.
+ */
+function getEventualUrl(a) {
+  const urlParamName = a.getAttribute('data-url-param-name') || 'adurl';
+  const eventualUrl = parseQueryString(a.search)[urlParamName];
+  if (!eventualUrl) {
+    return;
+  }
+  if (!eventualUrl.indexOf('https://cdn.ampproject.org/c/') == 0) {
+    return;
+  }
+  return eventualUrl;
+}
+
+/**
+ * Navigate to the given URL. Infers the target
+ * @param {!Window} win
+ * @param {!Element} a Anchor element
+ * @param {string} url
+ */
+function navigateTo(win, a, url) {
+  const target = (a.target || '_top').toLowerCase();
+  win.open(url, target);
+}
+
+/**
+ * Establishes a connection to the AMP Cache and makes sure
+ * the AMP JS is cached.
+ * @param {!Window} win
+ */
+export function warmupStatic(win) {
+  // Preconnect using an image, because that works on all browsers.
+  // The image has a 1 minute cache time to avoid duplicate
+  // preconnects.
+  new win.Image().src = 'https://cdn.ampproject.org/preconnect.gif';
+  // Preload the primary AMP JS that is render blocking.
+  const linkRel = /*OK*/document.createElement('link');
```

Comment:
It doesn't matter which document creates the element. I somehow prefer this style.